File: src/main/java/refactown/cleancode/c03functions/RockPaperScissors.java
package refactown.cleancode.c03functions;

import java.util.Arrays;
import java.util.List;

public class RockPaperScissors {
	
	private static List<Character> VALID_CHOOSES =  Arrays.asList('R', 'S', 'P');

	public static int rockPaperScissors(char first, char second) {
	    if (!VALID_CHOOSES.contains(first)) {
	        throw new IllegalArgumentException("Invalid first choose:" + first + ". Use " + VALID_CHOOSES);
	    }
	    if (!VALID_CHOOSES.contains(second)) {
	        throw new IllegalArgumentException("Invalid second choose:" + second + ". Use " + VALID_CHOOSES);
	    }
	    if (first == 'R') {
	        if (second == 'P') return 1; else if (second == 'S') return -1; else return 0;
	    } else if (first == 'P') {
	        if (second == 'S') return 1; else if (second == 'R') return -1; else return 0;
	    } else { // S
	        if (second == 'R') return 1; else if (second == 'P') return -1; else return 0;
	    }
	
	}
}

File: src/main/kotlin/refactown/cleancode/c03functions/RockPaperScissors.kt
package refactown.cleancode.c03functions

import java.lang.IllegalArgumentException

private val VALID_CHOICES = listOf('R', 'S', 'P') // Immutable

fun rockPaperScissors(first: Char, second: Char): Int {
    if (first !in VALID_CHOICES) {
        throw IllegalArgumentException("Invalid first choice: $first. Use $VALID_CHOICES")
    }
    if (second !in VALID_CHOICES) {
        throw IllegalArgumentException("Invalid second choice: $second. Use $VALID_CHOICES")
    }
    return if (first == 'R') {
        if (second == 'P') 1 else if (second == 'S') -1 else 0
    } else if (first == 'P') {
        if (second == 'S') 1 else if (second == 'R') -1 else 0
    } else { // S
        if (second == 'R') 1 else if (second == 'P') -1 else 0
    }
}
